Re-organization of A2rchi, Workflows, ChainWrapper; Prompt formatting and abstraction; CLI re-rewrite; Configuration Handling; Other misc. improvements

lucalavezzo opened the pull request:
### Class structure

This PR seeks to re-organize the classes that handle the chains. The new structure is:

- `A2rchi`: this global class is what is called by the interfaces, updates the vectorstore connection, and invokes a given Pipeline selected via the configuration.
- `Pipeline`: sequences of `langchain`'s chains, retrievers, etc. (previously `BaseChain`s). Each A2rchi instance would pick a `Pipeline` to operate. e.g. The basic Q/A is QAPipeline, which defines the set of operations that are called: first a condensing step of the history, then the retrieval of documents, and finally the answer with the documents in the context.
- `ChainWrapper`: a lightweight, generic class to wrap `langchain`'s chains, to harmonize them with our prompts and inputs.

### Prompt abstraction

Further, prompts are abstracted in the following ways:

- The order of the input variables is now completely flexible.
- Which input variables are used is flexible within each `Pipeline`; each `Pipeline` can define a set of required input variables.
- Prompts are fully defined via the config. Any prompt can be used in any`Pipeline`.
- Required/optional prompts for a given `Pipeline` configured via the config

### CLI

Complete re-write. Pietro might want to add more here: TODO

### Config updating in A2rchi

The config file is loaded in many parts of the codebase, previously statically, at script init, and not again.
For all parts concerning the `A2rchi` class, the config is loaded only through the class itself. It can be updated with `A2rchi.update()`. This allows tighter control over what config we are currently using, with the idea we can set up a system to switch between configs now.
Global and interfaces configs are still considered static, i.e. loaded at script init time, and not again.

### Other

- `Cleo` -> `Redmine` everywhere
- `PromptLogger` now in a new utilities script
- `TokenLimiter` bug fix for the documents reduction, and expanded to cover any given input variable, which for now is left as unprunable. Whether an input is a list of documents is automatically determined.
- `{context}` -> `{retriever_output}` and `{condensed_question}` -> `{condensed_output}`. The idea being that soon we can move to a more generic MODEL_output.
- Chat shows top 5 links above a certain threshold; removes duplicate links

lucalavezzo commented:
Merge conflicts are fixed.

pmlugato commented on a2rchi/chains/workflows.py:
```diff
@@ -1,129 +1,137 @@
+        self.condense_prompt = PROMPTS["CONDENSING_PROMPT"]
+
+        # grab models
+        model_class_map = self.chain_config["MODEL_CLASS_MAP"]
+        model_name = self.chain_config.get("MODEL_NAME", None)
+        condense_model_name = self.chain_config.get("CONDENSE_MODEL_NAME", model_name)
+        self.llm = model_class_map[model_name]["class"](**model_class_map[model_name]["kwargs"])
```
One question here: in this case, if we are deploying both the chatbot and the Redmine/Mailer services (or any two services), with different configurations -- e.g., different LLMs -- would the logic here be to write a workflow for each service, and there is hardcoded  `model_name = self.chain_config.get("REDMINE_MODEL_NAME")`? Or is this workflow meant to be adaptive in some way?

Not against it, if anything could force us to be a bit more intentional in differentiating the two... and also probably more straightforward to maintain, even if all of the logic is the same. Not immediately sure what is best.

lucalavezzo commented:
This pull request introduces a major refactor of the core pipeline and chain architecture for the A2rchi framework. The changes focus on modularizing pipelines, improving configuration and prompt handling, introducing a generic chain wrapper, and cleaning up legacy code. The most important updates are grouped below:

### Pipeline Architecture & Refactor

* Introduced a new `BasePipeline` and `QAPipeline` class in `a2rchi/chains/pipelines.py` (renamed from `base.py`), replacing the previous `BaseQAChain` with a more flexible, modular, and configuration-driven pipeline structure. Pipelines now handle prompt/model initialization, input preparation, and retriever updates in a cleaner, extensible way.
* Refactored pipeline invocation logic: the `QAPipeline` class now manages prompt selection, LLM/model instantiation, input variable handling, and document retrieval, providing a more robust and maintainable workflow.

### Chain Wrappers & Utilities

* Added a new `ChainWrapper` class in `a2rchi/chains/chain_wrappers.py` to standardize how LangChain chains are wrapped, including prompt validation, token limiting, and input preparation. This improves input/output consistency and error handling across chains.
* Moved and improved prompt logging and token limiting utilities to dedicated modules, further decoupling concerns and improving readability. [[1]](diffhunk://#diff-f81c9be52d9a78467a1312cafffbd570ac8ed8ab75f7c913bdc31cba0940731dR1-R96) [[2]](diffhunk://#diff-7e3486dbc314425f57069fe4397416f496b140acc4142558e531f5ec485454bfL4-R215)

### Configuration & Prompt Handling

* Enhanced configuration-driven initialization: models, prompts, and other components are now loaded and validated based on config files, with improved logging of model parameters via the new `print_model_params` utility. [[1]](diffhunk://#diff-4b4fc58ab80d34a91a7c3ccfa85140baecb67df23c0cda9bf381d2dcf5923940R27-R31) [[2]](diffhunk://#diff-7e3486dbc314425f57069fe4397416f496b140acc4142558e531f5ec485454bfL4-R215)
* Prompts are now validated and loaded using `ValidatedPromptTemplate` and `read_prompt`, ensuring that all required input variables are present and reducing runtime errors.

### Central Framework Class

* Added a new `A2rchi` class in `a2rchi/chains/a2rchi.py` that acts as the central entry point for the framework, managing configuration loading, pipeline instantiation, and vectorstore connections. This class simplifies usage and enforces consistent initialization and invocation patterns.

### Cleanup & Legacy Code Removal

* Removed legacy and redundant code, such as the old `PromptLogger` class and unused methods, to streamline the codebase and avoid confusion.

These changes collectively modernize the A2rchi framework, making it more modular, maintainable, and easier to extend for new pipelines and use cases.

**References:**  
[[1]](diffhunk://#diff-7e3486dbc314425f57069fe4397416f496b140acc4142558e531f5ec485454bfL4-R215) [[2]](diffhunk://#diff-f81c9be52d9a78467a1312cafffbd570ac8ed8ab75f7c913bdc31cba0940731dR1-R96) [[3]](diffhunk://#diff-d06223068220eff8797c454a247c1cc7dbf0ee5d12b8128f0c9bfa5eb6de112fR1-R160) [[4]](diffhunk://#diff-4b4fc58ab80d34a91a7c3ccfa85140baecb67df23c0cda9bf381d2dcf5923940R27-R31) [[5]](diffhunk://#diff-7e3486dbc314425f57069fe4397416f496b140acc4142558e531f5ec485454bfL345-L383)
